source/common/http/conn_manager_impl.cc

```diff
@@ -1745,16 +1750,30 @@ bool ConnectionManagerImpl::ActiveStream::createFilterChain() {
       state_.successful_upgrade_ = true;
       connection_manager_.stats_.named_.downstream_cx_upgrades_total_.inc();
       connection_manager_.stats_.named_.downstream_cx_upgrades_active_.inc();
-      return true;
+      return CreateFilterChainResult::UPGRADED;
     } else {
-      upgrade_rejected = true;
-      // Fall through to the default filter chain. The function calling this
-      // will send a local reply indicating that the upgrade failed.
+      if (absl::EqualsIgnoreCase(request_headers_->Upgrade()->value().getStringView(),
+                                 Http::Headers::get().UpgradeValues.H2c)) {
+        request_headers_->removeUpgrade();
+        upgrade_ignored = true;
+      } else {
+        upgrade_rejected = true;
+        // Fall through to the default filter chain. The function calling this
+        // will send a local reply indicating that the upgrade failed.
+      }
     }
   }
 
   connection_manager_.config_.filterFactory().createFilterChain(*this);
-  return !upgrade_rejected;
+  if (upgrade_rejected) {
```

Comment:
That's a scary flow...
3 flags generates 4 conditions. 
Could you move it to `  if (upgrade != nullptr) {` block?

Reply:
done.